src/machine/machine-varlink.c

```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
+        if (r == -ESRCH)
+                return sd_varlink_error(link, "io.systemd.Machine.NoSuchMachine", NULL);
+        if (r != 0)
+                return r;
+
+        r = varlink_verify_polkit_async(
+                        link,
+                        manager->bus,
+                        machine_open_polkit_action(p.mode, machine->class),
+                        machine_open_polkit_details(p.mode, machine->name, user, path, command_line),
+                        &manager->polkit_registry);
+        if (r <= 0)
+                return r;
+
+        master = machine_openpt(machine, O_RDWR|O_NOCTTY|O_CLOEXEC, &pty_name);
+        if (ERRNO_IS_NEG_NOT_SUPPORTED(master))
+                return sd_varlink_error(link, "io.systemd.Machine.NotSupported", NULL);
+        if (master < 0)
+                return log_debug_errno(master, "Failed to open pseudo terminal: %m");
+
+        switch (p.mode) {
+                case MACHINE_OPEN_MODE_LOGIN:
+                        r = machine_start_getty(machine, pty_name, /* error = */ NULL);
+                        if (r == -ENOENT)
+                                return sd_varlink_error(link, "io.systemd.Machine.NoSDBus", NULL);
+                        if (ERRNO_IS_NEG_NOT_SUPPORTED(r))
+                                return sd_varlink_error(link, "io.systemd.Machine.NotSupported", NULL);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to start getty for machine '%s': %m", machine->name);
+
+                        break;
+
+                case MACHINE_OPEN_MODE_SHELL: {
+                        _cleanup_close_ int slave = -EBADF;
+
+                        slave = machine_open_terminal(machine, pty_name, O_RDWR|O_NOCTTY|O_CLOEXEC);
+                        if (slave < 0)
+                                return log_debug_errno(slave, "Failed to open terminal: %m");
+
+                        assert(user && path && args); /* to avoid gcc complaining about possible uninitialized variables */
+                        r = machine_start_shell(machine, slave, pty_name, user, path, args, p.env, /* error = */ NULL);
+                        if (r == -ENOENT)
+                                return sd_varlink_error(link, "io.systemd.Machine.NoSDBus", NULL);
+                        if (ERRNO_IS_NEG_NOT_SUPPORTED(r))
+                                return sd_varlink_error(link, "io.systemd.Machine.NotSupported", NULL);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to start shell for machine '%s': %m", machine->name);
+
+                        slave = safe_close(slave);
+                        break;
+                }
+
+                default:
+                        break;
+        }
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_UNSIGNED("ptyFileDescriptor", master),
```

Comment:
to pass fds over varlink use sd_varlink_push_fd() and store the index integer it returns in the json.

Reply:
what do I do if `allow_fd_passing_output` is not enabled? just pass only name?

Reply:
I think I found answer: I need to enable it via `sd_varlink_set_allow_fd_passing_output()`